pieces/shared/menu-button/container.tsx

```diff
@@ -0,0 +1,170 @@
+import {
+  KeyboardEventHandler,
+  MouseEventHandler,
+  ReactNode,
+  useCallback,
+  useEffect,
+  useReducer,
+  useRef,
+  useState,
+} from 'react';
+
+import { css } from '@emotion/react';
+
+import Trigger from './trigger';
+import Menu, { MenuHandle } from './menu';
+
+const containerStyles = css({
+  display: 'flex',
+  position: 'relative',
+});
+
+const useInitialRender = () => {
+  const isInitialRenderRef = useRef(true);
+  useEffect(() => {
+    isInitialRenderRef.current = false;
+  }, []);
+  return { isInitialRender: isInitialRenderRef.current };
+};
+
+export const MenuButton = ({ label, children }: { label: string; children: ReactNode }) => {
+  const [isOpen, toggleIsOpen] = useReducer((isOpen) => !isOpen, false);
+
+  const [initialFocus, setInitialFocus] = useState<'first' | 'last'>('first');
+
+  const menuRef = useRef<MenuHandle>(null);
+
+  const triggerRef = useRef<HTMLButtonElement>(null);
+
+  const shouldResetFocusRef = useRef(false);
+
+  useEffect(() => {
+    if (!isOpen && shouldResetFocusRef.current) {
+      shouldResetFocusRef.current = false;
+      window.setTimeout(() => {
+        triggerRef.current?.focus();
+      }, 0);
+    }
+  }, [isOpen]);
+
+  const onKeyDown: KeyboardEventHandler<HTMLButtonElement> = useCallback((event) => {
+    switch (event.key) {
+      case 'ArrowDown':
+        event.preventDefault();
+        toggleIsOpen();
+        setInitialFocus('first');
+        break;
+
+      case 'ArrowUp':
+        event.preventDefault();
+        toggleIsOpen();
+        setInitialFocus('last');
+        break;
+    }
+  }, []);
+
+  useEffect(() => {
+    if (isOpen === false) {
+      setInitialFocus('first');
+      return;
+    }
+
+    if (initialFocus === 'first') {
+      menuRef.current?.focusFirst();
+    }
+
+    if (initialFocus === 'last') {
+      menuRef.current?.focusLast();
+    }
+  }, [initialFocus, isOpen]);
+
+  const containerRef = useRef<HTMLSpanElement>(null);
+
+  const onMenuKeyDown: KeyboardEventHandler<HTMLButtonElement> = useCallback((event) => {
+    switch (event.key) {
+      case 'Escape':
+        shouldResetFocusRef.current = true;
+        toggleIsOpen();
+        return;
+
+      case 'ArrowDown':
+        menuRef.current?.focusNext();
+        event.preventDefault();
+        return;
+
+      case 'ArrowUp':
+        menuRef.current?.focusPrev();
+        event.preventDefault();
+        break;
+
+      case 'Home':
+        menuRef.current?.focusFirst();
+        break;
+
+      case 'End':
+        menuRef.current?.focusLast();
+        break;
+    }
+
+    if (containerRef.current && /[a-z]/i.test(event.key)) {
```

Comment:
Q: can you explain what this is achieving?

Reply:
added a comment in the code, but this will select the next menu item which starts with the letter pressed